Compile project and run tests on CI

r-ash opened the pull request:
This PR will
* Run automated tests as part of CI on windows, linux and mac

This also sets some compiler flags via CMakePresets which turn on a few things. Got this set up from recommended https://github.com/friendlyanon/cmake-init/ for header only libraries. They look quite sensible, at the moment will only be used on CI. But it turns on a bunch of warnings, sets optimisation flags and source fortification and adds a bunch of security flags. You can use the presets locally using `cmake --preset ci-windows` or similar. We could add a preset for debug mode too.

I think it is worth taking a look at some of the warnings that are raised here and seeing if they are something we should fix.

r-ash committed:
Compile code and run tests on CI

r-ash committed:
Install boost before building GoalsARM

r-ash committed:
Add CMake presets

r-ash committed:
Set toolset and platform version explicitly

r-ash committed:
Fix presets

r-ash committed:
Use boost version which exists in prebuilt

r-ash committed:
Use ubuntu 22.04

r-ash committed:
Explicitly set boost path

r-ash committed:
Set CMP0144 to new policy

r-ash committed:
Install boost not into root dir

r-ash committed:
Turn off warnings on boost deps and fix floating point comparison errors

r-ash committed:
Also ignore build warnings from ctest

r-ash commented on CMakeLists.txt:
```diff
@@ -13,7 +16,7 @@ if(POLICY CMP0167)
 endif()
 find_package(Boost 1.82 REQUIRED)
-target_include_directories(GoalsARM INTERFACE ${Boost_INCLUDE_DIRS})
+target_include_directories(GoalsARM SYSTEM INTERFACE ${Boost_INCLUDE_DIRS})
```
This stops the compiler from raising warnings from boost source code

rlglaubius replied:
If this cuts down on the large number of warnings generated by partial template matches in boost then this is very welcome

r-ash replied:
Yep it should ignore all those now!

r-ash commented on tests/CMakeLists.txt:
```diff
@@ -28,6 +30,7 @@ foreach(TEST_FILE ${TEST_SOURCES})
     get_filename_component(TEST_NAME ${TEST_FILE} NAME_WE)
     file(RELATIVE_PATH FILE_NAME ${CMAKE_CURRENT_SOURCE_DIR} ${TEST_FILE})
     add_executable(${TEST_NAME} ${FILE_NAME})
+    target_include_directories(Catch2 SYSTEM INTERFACE ${catch2_includes})
```
This stops warnings being raised from catch2 source code

r-ash commented on include/DPProjection_impl.h:
```diff
@@ -919,8 +919,8 @@ namespace DP {
 		// Short-circuit uptake calculations if no one is on ART
-		if (dat.art_num_adult( t, DP::FEMALE) == 0.0 && dat.art_num_adult( t, DP::MALE) == 0.0 &&
-			  dat.art_prop_adult(t, DP::FEMALE) == 0.0 && dat.art_prop_adult(t, DP::MALE) == 0.0) {
+		if (GB::isZero(dat.art_num_adult( t, DP::FEMALE)) && GB::isZero(dat.art_num_adult( t, DP::MALE)) &&
```
One of the new flags was causing checking equality of floats with `==` or `!=` to raise an error. Because in many cases this is not what you want. But I think here it is? We are initialising these with doubles as `0.0` so the check was just checking that this had not changed from its initialised value, is that right? If so, seems like our check that it is exactly equal to 0.0 was correct in this case. I have moved this into a separate function and ignored the warnings for just that function (and added a note why)

rlglaubius replied:
Thanks for looking out for this. In this case the exact numerical match to initial value of 0.0 is intended. I do like having isZero in the codebase either way. Eventually we'll probably want a more general isEqual(value, threshold) as well, but apparently that has not come up yet (I've mostly been adding epsilon to denominators to minimize conditional logic in inner loops, which helps reduce the need for these predicates).

r-ash committed:
Remove return on void functions

r-ash commented on include/DPData.h:
```diff
@@ -195,10 +195,10 @@ namespace DP {
 
 		inline double pwid_infection_force(const int t, const int s) const {return (*_pwid_infection_force)[t][s];}
-		inline void pwid_infection_force(const int t, const int s, const double value) {return (*_pwid_infection_force)[t][s] = value;}
+		inline void pwid_infection_force(const int t, const int s, const double value) {(*_pwid_infection_force)[t][s] = value;}
```
This was causing errors downstream https://github.com/AvenirHealth-org/GoalsARMPython/actions/runs/13008790493/job/36281570892#step:6:89

rlglaubius replied:
Thanks for catching this. Agree with the change.

r-ash committed:
Fix small typo in comment

rlglaubius commented:
Do the VSCode integration steps need to be updated? These work for me in the `main` branch, but not in `ci`. I started from a fresh clone of GoalsARM and switched immediately to the `ci` branch, then opened the folder in VSCode. CMake automatically wrote build files. So far so good.

The first step is to select a kit for GoalsARM, but I do not see that item in the command palette. I don't see "CMake: Select variant" either. If I skip ahead and try to build (Ctrl+Shift+P > CMake: Build) I get the following error:

```
[main] Building folder: C:/Proj/Temp/GoalsARM/build 
[build] Starting build
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --build C:/Proj/Temp/GoalsARM/build --parallel 22 --
[build] MSBuild version 17.5.0+6f08c67f3 for .NET Framework
[build] 
[build] C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(438,5): error MSB8013: This project doesn't contain the Configuration and Platform combination of Debug|x64. [C:\Proj\Temp\GoalsARM\build\ZERO_CHECK.vcxproj]
[proc] The command: "C:\Program Files\CMake\bin\cmake.EXE" --build C:/Proj/Temp/GoalsARM/build --parallel 22 -- exited with code: 1
[driver] Build completed: 00:00:00.273
[build] Build finished with exit code 1
[rollbar] Unhandled exception: Unhandled Promise rejection: refreshTests Error: Build failed. {}
```

My best guess is that there's at least one additional step needed, but I'm not sure what that would be. In case it's helpful, here is the output I get when I point VSCode at the GoalsARM folder:

```
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --version
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" -E capabilities
[kit] Successfully loaded 5 kits from C:\Users\RobertGlaubius\AppData\Local\CMakeTools\cmake-tools-kits.json
[presetController] Reading and validating the presets "file C:\Proj\Temp\GoalsARM\CMakePresets.json"
[presetController] Successfully validated C:\Proj\Temp\GoalsARM\CMakePresets.json against presets schema
[presetController] Expanding presets file C:\Proj\Temp\GoalsARM\CMakePresets.json
[presetController] Successfully expanded presets file C:\Proj\Temp\GoalsARM\CMakePresets.json
[main] Configuring project: GoalsARM 
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" "-DCMAKE_CXX_FLAGS=/sdl /guard:cf /utf-8 /diagnostics:caret /w14165 /w44242 /w44254 /w44263 /w34265 /w34287 /w44296 /w44365 /w44388 /w44464 /w14545 /w14546 /w14547 /w14549 /w14555 /w34619 /w34640 /w24826 /w14905 /w14906 /w14928 /w45038 /W4 /permissive- /volatile:iso /Zc:inline /Zc:preprocessor /Zc:enumTypes /Zc:lambda /Zc:__cplusplus /Zc:externConstexpr /Zc:throwingNew /EHsc" "-DCMAKE_EXE_LINKER_FLAGS=/machine:x64 /guard:cf" -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_CXX_STANDARD=20 -DCMAKE_CXX_STANDARD_REQUIRED=ON -Dheaderonly_DEVELOPER_MODE=ON -DCMAKE_CONFIGURATION_TYPES=Release -Wdev -Wdeprecated --warn-uninitialized -Werror=dev -Werror=deprecated -SC:/Proj/Temp/GoalsARM -BC:/Proj/Temp/GoalsARM/build -G "Visual Studio 17 2022" -A x64
[cmake] Warn about uninitialized values.
[cmake] -- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22631.
[cmake] -- Locating Catch2
[cmake] -- Could NOT find Catch2 (missing: Catch2_DIR)
[cmake] --   Catch2 not found, fetching dynamically...
[cmake] --   Catch2 fetched succesfully
[cmake] -- Configuring done (1.5s)
[cmake] -- Generating done (0.7s)
[cmake] -- Build files have been written to: C:/Proj/Temp/GoalsARM/build
```

r-ash committed:
Tidy up preset usage and document

r-ash committed:
Install Ninja on linux during CI

r-ash commented:
Thanks Rob for the thorough review! I forgot to check these on a fresh install. I've updated the "preset" configuration now and gone a bit deeper into it for local development. I think what the ends with is fairly nice. One bit of ugliness because the CMake extension in vscode is a bit out of date, but there is an easy work around. I've updated the docs with usage instructions, let me know how you get on with those. Happy to chat through these presets also if helpful.

r-ash committed:
Add fallback user presets
